File: app/src/main/java/com/kotobyte/search/SearchPageContracts.java
package com.kotobyte.search;

import com.kotobyte.models.Kanji;
import com.kotobyte.models.Word;

import java.util.List;

import io.reactivex.Single;

interface SearchPageContracts {

    interface View {

        void showWordSearchProgressBar(boolean show);
        void showWordSearchResultsView(boolean show);
        void showWordSearchResults(List<Word> words);
        void showNoWordSearchResultsLabel(boolean show);
        void showKanjiSearchResults(int position, List<Kanji> kanjiList);
        void showKanjiDetailScreen(Kanji kanji);
        void showUnknownError(Throwable error);
    }

    interface Presenter {

        void onCreate();
        void onDestroy();

        void onRequestKanjiListForWord(int position, Word word);
        void onRequestDetailForKanji(Kanji kanji);
    }

    interface DataSource {
        Single<List<Word>> searchWords(String query);
        Single<List<Kanji>> searchKanji(String query);
    }
}

File: app/src/main/java/com/kotobyte/search/SearchPageContracts.kt
package com.kotobyte.search

import com.kotobyte.models.Kanji
import com.kotobyte.models.Word

internal object SearchPageContracts {

    interface View {

        fun showWordSearchProgressBar(show: Boolean)
        fun showWordSearchResultsView(show: Boolean)
        fun showWordSearchResults(words: List<Word>)
        fun showNoWordSearchResultsLabel(show: Boolean)
        fun showKanjiSearchResults(position: Int, kanjiList: List<Kanji>)
        fun showKanjiDetailScreen(kanji: Kanji)
        fun showUnknownError(error: Throwable)
    }

    interface Presenter {

        fun onCreate()
        fun onDestroy()

        fun onRequestKanjiListForWord(position: Int, word: Word)
        fun onRequestDetailForKanji(kanji: Kanji)
    }

    interface DataSource {
        fun searchWords(query: String): List<Word>
        fun searchKanji(query: String): List<Kanji>
    }
}
